kratos/solving_strategies/convergencecriterias/mixed_generic_criteria.h

```diff
@@ -0,0 +1,441 @@
+     * Note that these arrays are assumed to be already initialized to zero.
+     * @param rModelPart Reference to the ModelPart containing the fluid problem.
+     * @param rDofSet Reference to the container of the problem's degrees of freedom (stored by the BuilderAndSolver)
+     * @param rDx Vector of results (variations on nodal variables)
+     * @param rDofsCount Array containing the number of DOFs per variable
+     * @param rSolutionNormsVector Array containing the solution norms accumulated values for each variable checked
+     * @param rIncreaseNormsVector Array containing the correction norms accumulated values for each variable checked
+     */
+    virtual void GetNormValues(
+        const ModelPart& rModelPart,
+        const DofsArrayType& rDofSet,
+        const TSystemVectorType& rDx,
+        std::vector<int>& rDofsCount,
+        std::vector<TDataType>& rSolutionNormsVector,
+        std::vector<TDataType>& rIncreaseNormsVector)
+    {
+        int n_dofs = rDofSet.size();
+
+        // Loop over Dofs
+#pragma omp parallel
+        {
+            // Local thread variables
+            int dof_id;
+            TDataType dof_dx;
+            TDataType dof_value;
+
+            // Local reduction variables
+            std::vector<TDataType> var_solution_norm_reduction(mVariableSize);
+            std::vector<TDataType> var_correction_norm_reduction(mVariableSize);
+            std::vector<int> dofs_counter_reduction(mVariableSize);
+            for (int i = 0; i < mVariableSize; i++) {
+                var_solution_norm_reduction[i] = 0.0;
+                var_correction_norm_reduction[i] = 0.0;
+                dofs_counter_reduction[i] = 0;
+            }
+
+#pragma omp for
+            for (int i = 0; i < n_dofs; i++) {
+                auto it_dof = rDofSet.begin() + i;
+                if (it_dof->IsFree()) {
+                    dof_id = it_dof->EquationId();
+                    dof_value = it_dof->GetSolutionStepValue(0);
+                    dof_dx = TSparseSpace::GetValue(rDx, dof_id);
+
+                    const auto &r_current_variable = it_dof->GetVariable();
+                    int var_local_key = mLocalKeyMap[r_current_variable.IsComponent() ? r_current_variable.GetSourceVariable().Key() : r_current_variable.Key()];
+
+                    var_solution_norm_reduction[var_local_key] += dof_value * dof_value;
+                    var_correction_norm_reduction[var_local_key] += dof_dx * dof_dx;
+                    dofs_counter_reduction[var_local_key]++;
+                }
+            }
+
+#pragma omp critical
+            {
+                for (int i = 0; i < mVariableSize; i++) {
+                    rDofsCount[i] += dofs_counter_reduction[i];
+                    rSolutionNormsVector[i] += var_solution_norm_reduction[i];
+                    rIncreaseNormsVector[i] += var_correction_norm_reduction[i];
+                }
+            }
+        }
+    }
```

Comment:
you could already now do this with the parallel utils by using the reducer (just saying)

Reply:
Let's do it in a future PR. @sunethwarna is waiting for this to clean up the RANS PRs.